libraries/Networking/src/NetworkManager.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "NetworkManager.h"
+#include "NetworkInterface.h"
+#include "esp_netif.h"
+#include "lwip/ip_addr.h"
+#include "lwip/dns.h"
+#include "esp32-hal-log.h"
+#include "esp_mac.h"
+
+NetworkManager::NetworkManager(){
+
+}
+
+bool NetworkManager::begin(){
+    static bool initialized = false;
+    if(!initialized){
+        initialized = true;
+#if CONFIG_IDF_TARGET_ESP32
+        uint8_t mac[8];
+        if(esp_efuse_mac_get_default(mac) == ESP_OK){
+            esp_base_mac_addr_set(mac);
+        }
+#endif
+        initialized = esp_netif_init() == ESP_OK;
+        if(!initialized){
+        	log_e("esp_netif_init failed!");
+        }
+    }
+    if(initialized){
+    	initNetworkEvents();
+    }
+    return initialized;
+}
+
+typedef struct gethostbynameParameters {
+    const char *hostname;
+    ip_addr_t addr;
+    uint8_t addr_type;
+    int result;
+} gethostbynameParameters_t;
+
+/**
+ * DNS callback
+ * @param name
+ * @param ipaddr
+ * @param callback_arg
+ */
+static void wifi_dns_found_callback(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    gethostbynameParameters_t *parameters = static_cast<gethostbynameParameters_t *>(callback_arg);
+    if(ipaddr) {
+        if(parameters->result == 0){
+            memcpy(&(parameters->addr), ipaddr, sizeof(ip_addr_t));
+            parameters->result = 1;
+        }
+    } else {
+        parameters->result = -1;
+    }
+    Network.setStatusBits(NET_DNS_DONE_BIT);
+}
+
+/**
+ * Callback to execute dns_gethostbyname in lwIP's TCP/IP context
+ * @param param Parameters for dns_gethostbyname call
+ */
+static esp_err_t wifi_gethostbyname_tcpip_ctx(void *param)
+{
+    gethostbynameParameters_t *parameters = static_cast<gethostbynameParameters_t *>(param);
+    return dns_gethostbyname_addrtype(parameters->hostname, &parameters->addr, &wifi_dns_found_callback, parameters, parameters->addr_type);
+}
+
+/**
+ * Resolve the given hostname to an IP address.
+ * @param aHostname     Name to be resolved
+ * @param aResult       IPAddress structure to store the returned IP address
+ * @return 1 if aIPAddrString was successfully converted to an IP address,
+ *          else error code
+ */
+int NetworkManager::hostByName(const char* aHostname, IPAddress& aResult, bool preferV6)
+{
+    err_t err = ERR_OK;
+    gethostbynameParameters_t params;
+
+    // This should generally check if we have a global address assigned to one of the interfaces.
+    // If such address is not assigned, there is no point in trying to get V6 from DNS as we will not be able to reach it.
+    // That is of course, if 'preferV6' is not set to true
```

Comment:
NOTE: This should work in most cases; I will try and test the different scenario combinations (e.g. v4-only vs v6-only vs dual stack networks and destinations), for host lookup.

However when I was exploring ESP-IDF while I got this approach to work for HTTP calls, it did not work for HTTPS -- for HTTPS I had to pass in the hostname (needed to validate certificate) and the TLS component did the lookup, which I had no control over.

Internally it used lwip_getaddrinfo(), which at best does V4_V6 preference order. 

I will also try and test HTTPS scenarios.

I have a PR up for ESP-LWIP to implement RFC 6724 destination address ordering, based on available source addresses. If Arduino-ESP32 has a similar problem them it will need to wait for the upstream fix. (But I haven't checked yet).